Add Segment Anything Model (SAM)

ArthurZucker opened the pull request:
# What does this PR do?

Original repo: https://github.com/facebookresearch/segment-anything

Segment Anything Model (SAM) is a recent model from Meta AI that makes it possible to predict image segmentation masks given an image and various inputs such as bounding boxes, 2D points or previous masks.
It is also mentioned in the original paper that the model can take textual input, but this feature has not been released yet in the original repository.
The release came with 3 weights, namely:

- `sam_vit_b`
- `sam_vit_h`
- `sam_vit_l`

Their main difference is about the vision encoder size, the prompt encoder, and mask decoder should stay the same.

According to the paper, for each input, the model predicts 3 binary masks, corresponding to the region where the "object of interest" lives in the image.

cc @sgugger @amyeroberts

ArthurZucker committed:
initial commit

younesbelkada committed:
keys match

younesbelkada committed:
update, fix conversion

younesbelkada committed:
fixes, inference working

younesbelkada committed:
fix

younesbelkada committed:
more fixes

younesbelkada committed:
more fixes

younesbelkada committed:
clean up

younesbelkada committed:
more clean up

ArthurZucker commented on src/transformers/models/sam/modeling_sam.py:
```diff
@@ -53,268 +55,1039 @@
-        )
+# From https://github.com/facebookresearch/detectron2/blob/main/detectron2/layers/batch_norm.py # noqa
+# Itself from https://github.com/facebookresearch/ConvNeXt/blob/d1fa8f6fef0a165b27399986cc2bdacc92777e40/models/convnext.py#L119  # noqa
+class LayerNorm2d(nn.Module):
```
Already implemented in transformers : see [here](https://github.com/ArthurZucker/transformers/blob/f36ef7df37d7017df1fe889fd64382844af43bac/src/transformers/models/convnext/modeling_convnext.py#L98-L126)

ArthurZucker commented on src/transformers/models/sam/modeling_sam.py:
```diff
@@ -53,268 +55,1039 @@
-# Copied from transformers.models.blip.modeling_blip.BlipVisionEmbeddings with Blip->Sam
-class SamVisionEmbeddings(nn.Module):
-    def __init__(self, config: SamVisionConfig):
+class TwoWayTransformer(nn.Module):
```
If an embedding layer exists, it should be places before this

ArthurZucker commented on src/transformers/models/sam/modeling_sam.py:
```diff
@@ -53,268 +55,1039 @@
+            scale * torch.randn((2, num_pos_feats)),
+        )
 
-# Copied from transformers.models.blip.modeling_blip.BlipMLP
```
could have been kept and wrapped in the FF layer if there are multiple mlps

ArthurZucker committed:
fix copies and add convext copied layer norm

ArthurZucker committed:
stash

ArthurZucker committed:
pretty big upfate

ArthurZucker committed:
cleaning

ArthurZucker committed:
more cleaning

ArthurZucker committed:
fixup stuffs

ArthurZucker committed:
fix copies

ArthurZucker committed:
fix iinit

ArthurZucker committed:
update test removing tokenizer

ArthurZucker committed:
nits

ArthurZucker committed:
add pretrained

ArthurZucker committed:
more nits

ArthurZucker committed:
remove tracking of pipeline

younesbelkada committed:
few fixes

ArthurZucker committed:
update san and conversion script

younesbelkada committed:
fix mask decoder and prompt encoder conversion

younesbelkada committed:
Merge branch 'add-sam' of https://github.com/ArthurZucker/transformers into HEAD

younesbelkada committed:
fixes

ArthurZucker committed:
small update

ArthurZucker committed:
Merge branch 'add-sam' of https://github.com/ArthurZucker/transformers into add-sam

ArthurZucker committed:
fix order

younesbelkada committed:
fix

younesbelkada committed:
Merge branch 'add-sam' of https://github.com/ArthurZucker/transformers into HEAD

younesbelkada committed:
fix image embeddings

ArthurZucker committed:
nites

ArthurZucker committed:
Merge branch 'add-sam' of https://github.com/ArthurZucker/transformers into add-sam

younesbelkada committed:
few fixes

younesbelkada committed:
fix logits

younesbelkada committed:
clean up

younesbelkada committed:
fixes boxes inference

younesbelkada committed:
v1 AMG

younesbelkada committed:
clean up

younesbelkada committed:
some clean up

younesbelkada committed:
multi points support

younesbelkada committed:
amg working

younesbelkada committed:
fixup

younesbelkada committed:
clean up

younesbelkada committed:
readme

younesbelkada committed:
update toctree

younesbelkada committed:
fix type hint

younesbelkada committed:
multiple fixes

younesbelkada committed:
fixup

younesbelkada committed:
fixes

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_22654). All of your documentation changes will be reflected on that endpoint.

younesbelkada committed:
updates

younesbelkada committed:
updates

younesbelkada committed:
more tests

younesbelkada committed:
few fixes

younesbelkada committed:
Merge remote-tracking branch 'upstream/main' into HEAD

younesbelkada committed:
change to `SamForMaskGeneration`

younesbelkada committed:
doc

ArthurZucker committed:
fixup

ArthurZucker committed:
Merge branch 'add-sam' of https://github.com/ArthurZucker/transformers into add-sam

younesbelkada committed:
fix more tests

younesbelkada committed:
multiple fixes

younesbelkada committed:
fix CI tests

younesbelkada committed:
refactor processor

ArthurZucker committed:
renamings

ArthurZucker committed:
Merge branch 'add-sam' of https://github.com/ArthurZucker/transformers into add-sam

ArthurZucker committed:
draft the pipeline

younesbelkada committed:
refactor

younesbelkada committed:
Merge branch 'add-sam' of https://github.com/ArthurZucker/transformers into HEAD

younesbelkada committed:
fix tests

ArthurZucker committed:
Merge branch 'add-sam' of https://github.com/ArthurZucker/transformers into add-sam

younesbelkada committed:
fix test

younesbelkada committed:
few cleanings

younesbelkada committed:
fix test

ArthurZucker committed:
edit pipelien support chunking

ArthurZucker committed:
Merge branch 'add-sam' of https://github.com/ArthurZucker/transformers into add-sam

ArthurZucker committed:
udate

younesbelkada committed:
add slow tests

younesbelkada committed:
fix nit

younesbelkada committed:
fixup

younesbelkada committed:
fix nit

ArthurZucker committed:
current chunk pipleine

ArthurZucker committed:
Merge branch 'add-sam' of https://github.com/ArthurZucker/transformers into add-sam

ArthurZucker commented on src/transformers/pipelines/automatic_mask_generation.py:
```diff
@@ -0,0 +1,467 @@
+        total_points_per_crop.append(points_grid[layer_idxs[i]] * points_scale)
+
+    normalized_total_points_per_crop = []
+    # for points_per_crop in total_points_per_crop:
+    #     normalized_total_points_per_crop.append(
+    #         [normalize_coordinates(point, original_size) for point in points_per_crop]
+    #     )
+    # normaliwation should happen inside the processor
```
don't forget that I removed that

ArthurZucker commented on src/transformers/pipelines/automatic_mask_generation.py:
```diff
@@ -0,0 +1,467 @@
+    # normaliwation should happen inside the processor
+
+    if return_tensors == "pt":
+        import torch
```
```suggestion
```

ArthurZucker commented on src/transformers/pipelines/automatic_mask_generation.py:
```diff
@@ -0,0 +1,467 @@
+        image = load_image(image)
+        original_size = image.shape
+        crop_boxes, points_per_crop, cropped_images = generate_crop_boxes(image, **kwargs)
+        model_inputs = self.processor(images=cropped_images, return_tensors="pt").to("cuda")
```
points should be normalized here, so feed them as input

younesbelkada committed:
cast boxes in fp32

younesbelkada committed:
Merge branch 'add-sam' of https://github.com/ArthurZucker/transformers into HEAD

younesbelkada committed:
nit

ArthurZucker committed:
current updates

ArthurZucker committed:
piepleine works

ArthurZucker committed:
Merge branch 'add-sam' of https://github.com/ArthurZucker/transformers into add-sam

ArthurZucker committed:
fixup

younesbelkada committed:
clean up config

younesbelkada commented on docs/source/en/model_doc/sam.mdx:
```diff
@@ -0,0 +1,72 @@
+- According to the paper, textual input should be also supported. However, at this time of writing this seems to be not supported according to [the official repository](https://github.com/facebookresearch/segment-anything/issues/4#issuecomment-1497626844). 
+
+This model was contributed by [ybelkada](https://huggingface.co/ybelkada) and [ArthurZ](https://huggingface.co/ArthurZ).
+
```
@ArthurZucker add a mention of the pipeline here

ArthurZucker replied:
when the pipeline is merged!

younesbelkada committed:
fix slow tests

younesbelkada committed:
fix slow tests

younesbelkada committed:
clean up

ArthurZucker committed:
update doc and pipeline

ArthurZucker committed:
Merge branch 'add-sam' of https://github.com/ArthurZucker/transformers into add-sam

younesbelkada committed:
adds more slow tests

younesbelkada committed:
Merge branch 'add-sam' of https://github.com/ArthurZucker/transformers into HEAD

younesbelkada committed:
fix slow tests

ArthurZucker committed:
cleaning

ArthurZucker committed:
Merge branch 'add-sam' of https://github.com/ArthurZucker/transformers into add-sam

younesbelkada committed:
tests pass

younesbelkada committed:
add docstring

younesbelkada committed:
fix copies

younesbelkada committed:
clean up

ArthurZucker committed:
support batch of images

ArthurZucker committed:
Merge branch 'add-sam' of https://github.com/ArthurZucker/transformers into add-sam

ArthurZucker committed:
style

ArthurZucker committed:
dummy is needed, add tests

younesbelkada committed:
fix slow tests

younesbelkada committed:
Merge branch 'add-sam' of https://github.com/ArthurZucker/transformers into HEAD

younesbelkada committed:
fix CI

ArthurZucker committed:
update

ArthurZucker committed:
Merge branch 'add-sam' of https://github.com/ArthurZucker/transformers into add-sam

younesbelkada committed:
adds more tests

younesbelkada committed:
Merge branch 'add-sam' of https://github.com/ArthurZucker/transformers into HEAD

younesbelkada committed:
fixes

younesbelkada committed:
fixes

younesbelkada committed:
fixup

younesbelkada committed:
fixes

younesbelkada committed:
few fixes

younesbelkada committed:
filter

younesbelkada committed:
few fixes

younesbelkada committed:
some refactor

younesbelkada committed:
touches finales

younesbelkada committed:
fix

ArthurZucker committed:
style

ArthurZucker committed:
Merge branch 'add-sam' of https://github.com/ArthurZucker/transformers into add-sam

ArthurZucker committed:
remove pipeline files

younesbelkada committed:
fixes nits

ArthurZucker committed:
revert pipeline changes

ArthurZucker committed:
Merge branch 'add-sam' of https://github.com/ArthurZucker/transformers into add-sam

younesbelkada committed:
fix test

ArthurZucker committed:
fixup

ArthurZucker committed:
Merge branch 'add-sam' of https://github.com/ArthurZucker/transformers into add-sam

ArthurZucker commented on src/transformers/models/auto/modeling_auto.py:
```diff
@@ -976,6 +976,13 @@
     ]
 )
 
+MODEL_FOR_AUTOMATIC_MASK_GENERATION_NAMES = OrderedDict(
+    [
+        # Automatic Mask Generation
+        ("sam", "SamForMaskGeneration"),
+    ]
+)
+
```
removing

ArthurZucker committed:
remove automodel for automatic mask generation

ArthurZucker committed:
fix failing torch tests

ArthurZucker commented:
Might be a few things here and there, but new pairs of eyes will help us fix fast. Pinging @amyeroberts and @sgugger for a review!

sgugger commented on docs/source/en/model_doc/sam.mdx:
```diff
@@ -0,0 +1,72 @@
+- Fine-tuning the model is not supported yet
+- According to the paper, textual input should be also supported. However, at this time of writing this seems to be not supported according to [the official repository](https://github.com/facebookresearch/segment-anything/issues/4#issuecomment-1497626844). 
+
+This model was contributed by [ybelkada](https://huggingface.co/ybelkada) and [ArthurZ](https://huggingface.co/ArthurZ).
```
Don't forget to point out where the original code is.

sgugger commented on src/transformers/models/sam/configuration_sam.py:
```diff
@@ -0,0 +1,334 @@
+        self.prompt_encoder_config = SamPromptEncoderConfig(**prompt_encoder_config)
+        self.mask_decoder_config = SamMaskDecoderConfig(**mask_decoder_config)
+
+        self.initializer_factor = 1.0
+        self.initializer_range = 0.02
```
Should those be parametrizable?

sgugger commented on src/transformers/models/sam/image_processing_sam.py:
```diff
@@ -0,0 +1,658 @@
+        crop_box_x0 = [int((crop_w - overlap) * i) for i in range(n_crops_per_side)]
+        crop_box_y0 = [int((crop_h - overlap) * i) for i in range(n_crops_per_side)]
+
+        # Crops in XYWH format
```
It would be more helpful for future contributors and the maintenance if the names were more explicit:
```suggestion
        # Crops in XYWH format (left, top, width, height)
```
then all x -> left, y->top, w->width and h->height

w and h are explicit enough but x and y could be the coordinates of the center or any other corner.

younesbelkada replied:
These functions and all the functions below have been removed as they are specific to the AMG pipeline. We will address that in https://github.com/huggingface/transformers/pull/22840

sgugger commented on src/transformers/models/sam/image_processing_sam.py:
```diff
@@ -0,0 +1,658 @@
+
+
+def _uncrop_masks(masks, crop_box: List[int], orig_h: int, orig_w: int):
+    x0, y0, x1, y1 = crop_box
```
Very confusing here. It seems x1 and y1 are width and height? But the usual format for a box given with two points coordinates is the left/top corner and bottom/right corner. Again please be more explicit in the variable names.

younesbelkada replied:
Same comment as here: https://github.com/huggingface/transformers/pull/22654#discussion_r1171043296
But let's keep all that in mind for #22840 @ArthurZucker

sgugger commented on src/transformers/models/sam/image_processing_sam.py:
```diff
@@ -0,0 +1,658 @@
+
+def _batched_mask_to_box(masks):
+    """
+    Calculates boxes in XYXY format around masks. Return [0,0,0,0] for an empty mask. For input shape C1xC2x...xHxW,
```
What does XYXY mean? What does C1xC2x... mean?

sgugger commented on src/transformers/models/sam/image_processing_sam.py:
```diff
@@ -0,0 +1,658 @@
+    if torch.numel(masks) == 0:
+        return torch.zeros(*masks.shape[:-2], 4, device=masks.device)
+
+    # Normalize shape to CxHxW
```
No abbreviations in comments please, it defeats the purpose of the comment.

sgugger commented on src/transformers/models/sam/image_processing_sam.py:
```diff
@@ -0,0 +1,658 @@
+        masks = masks.unsqueeze(0)
+
+    # Get top and bottom edges
+    in_height, _ = torch.max(masks, dim=-1)
```
What does the in mean here?

sgugger commented on src/transformers/models/sam/image_processing_sam.py:
```diff
@@ -0,0 +1,658 @@
+    for count in rle["counts"]:
+        mask[idx : idx + count] = parity
+        idx += count
+        parity ^= True
```
```suggestion
        parity = not parity
```
Let's use something anyone will understand instead of c operators.

sgugger commented on src/transformers/models/sam/modeling_sam.py:
```diff
@@ -0,0 +1,1337 @@
+_CHECKPOINT_FOR_DOC = "facebook/sam-vit-h"
+
+SAM_PRETRAINED_MODEL_ARCHIVE_LIST = [
+    "facebook/sam-vit-h",
+    "facebook/sam-vit-l",
+    "facebook/sam-vit-b",
```
Can we use more explicit names? h/l/b is not very informative.

sgugger commented on src/transformers/models/sam/modeling_sam.py:
```diff
@@ -0,0 +1,1337 @@
+            Pixel values. Pixel values can be obtained using [`SamProcessor`]. See [`SamProcessor.__call__`] for
+            details.
+        input_points (`torch.FloatTensor` of shape `(batch_size, num_points, 2)`):
+            Input 2D spatial points, this is used by the prompt encoder to encode the prompt. Generally yields to much
+            better results. The points can be obtained by passing a list of list of list, corresponding to the batch,
+            the number of points per image and the x and y coordinate of the point. If a batch of images is passed,
+            "PAD" points are automatically generated and ignored by the prompt encoder.
```
This docstring is very unclear. The type annotation says this should be a tensor so why are you talking about a list of list. Is the list of lists what you need to pass the processor?
Also do the point coordinates need to be normalized between specific value (0 and ? -1 and 1?) or not.? How do you do a "PAD" point?

younesbelkada replied:
I should have addressed all the comments related to this docstring in [8440530](https://github.com/huggingface/transformers/pull/22654/commits/8440530e948e09bf3140176d4eea3f1136dcae1b) !

sgugger commented on src/transformers/models/sam/modeling_sam.py:
```diff
@@ -0,0 +1,1337 @@
+        input_labels (`torch.LongTensor` of shape `(batch_size, num_points)`):
+            Input labels for the points, this is used by the prompt encoder to encode the prompt. Generally yields to
+            according to the official implementation, there are 3 types of labels
+            - `1`: the point is a point that contains the object of interest
+            - `0`: the point is a point that does not contain the object of interest
+            - `-1`: the point corresponds to the background
```
List needs a new line before and after.

sgugger commented on src/transformers/models/sam/modeling_sam.py:
```diff
@@ -0,0 +1,1337 @@
+            - `-1`: the point corresponds to the background
+            We added the label:
+            - `-10`: the point is a padding point, thus should be ignored by the prompt encoder
+            The padding of the point is automatically done by the processor.
```
```suggestion
            The padding labels should be automatically done by the processor.
```

sgugger commented on src/transformers/models/sam/modeling_sam.py:
```diff
@@ -0,0 +1,1337 @@
+        input_boxes (`torch.FloatTensor` of shape `(batch_size, num_boxes, 4)`):
+            Input boxes for the points, this is used by the prompt encoder to encode the prompt. Generally yields to
+            much better results. The boxes can be obtained by passing a list of list of list, corresponding to the
+            batch, the number of boxes per image and the x1, y1, x2, y2 coordinates of the box.
```
What are x1, y1, x2 and y2? Be explicit.
Same comment as before on the list of list of list.

Here in the documentation of the model, explain clearly what the model expects.

sgugger commented on src/transformers/models/sam/modeling_sam.py:
```diff
@@ -0,0 +1,1337 @@
+            much better results. The boxes can be obtained by passing a list of list of list, corresponding to the
+            batch, the number of boxes per image and the x1, y1, x2, y2 coordinates of the box.
+        input_masks (`torch.FloatTensor` of shape `(batch_size, image_size, image_size)`):
+            Input masks for the points, this is used by the prompt encoder to encode the prompt. This can yield to
+            better results. The masks needs to be manually fed by the user.
```
Explain what this input should be :-)

sgugger commented on src/transformers/models/sam/modeling_sam.py:
```diff
@@ -0,0 +1,1337 @@
+            efficient computation, users can first retrieve the image embeddings using the `get_image_embeddings`
+            method, and then feed them to the `forward` method instead of feeding the `pixel_values`.
+        multimask_output (`bool`, *optional*):
+            Whether or not to return 3 masks instead of 1, as described in the original paper.
```
Could be expanded.

sgugger commented on src/transformers/models/sam/processing_sam.py:
```diff
@@ -0,0 +1,225 @@
+    single processor.
+
+    [`SamProcessor`] offers all the functionalities of [`SamImageProcessor`]. See the docstring of
+    [`~SamProcessor.__call__`] and [`~BlipProcessor.decode`] for more information.
```
The docstring of the call method of `SamProcessor` says nothing. Did you men `SamImageProcessor`? What's the link to Blip here?

sgugger reviewed (commented):
Thanks for your PR! I think more effort should go in the documentation (docstrings of the model, add examples in the docstring and doc page) before we merge.

Also for future maintenance, really avoid using x/y notations esp. when there are multiple points that (x, y) could represent (center, corner etc...)

amyeroberts reviewed (commented):
Thanks for all the work on getting this implemented so quickly! 🚀 

I've done a first, tried-to-catch-everything pass. 

Overall structure looks good (although I'm not 100% sure about not having a `SamModel` in the modeling file). 
Most comments are about consistency - namings of parameter deviate from other models and many of the docstrings are inconsistent, incomplete or wrong (looks like co-pilot generated issues and copy/pasta).
For the image processor, there's quite a lot of issues in the way it breaks design patterns and I won't accept it being merged as-is. I've [opened a PR](https://github.com/amyeroberts/transformers/pull/112) on my fork to show how it should be updated, in the interest of speed. The updates include: 
- Naming of methods and attributes to be consistent: postprocess -> post_process, target_sizes -> pad_size
- size parameters being dictionaries should be a dictionary
- Fixing docstrings (!)
- Fixing parameters not being used / correctly passed e.g. `size` was set but not used
- Remove breaking assumptions about the channel ordering of images
- Remove reimplementation of standard library methods
- Making sure all processing steps should be fully controllable from the processor call
- Image processor preprocessing methods should all accept and return numpy arrays
- Removal of unused methods (most of the private functions)
- All image processor methods should accept `data_format` as an argument

ArthurZucker committed:
update mdx

ArthurZucker committed:
revert removal of `MODEL_FOR_AUTOMATIC_MASK_GENERATION_MAPPING`

ArthurZucker committed:
update sam config based on review

Co-authored-by: amyeroberts <aeroberts4444@gmail.com>
Co-authored-by: sgugger <sylvain.gugger@gmail.com>

ArthurZucker committed:
update low_resolution_masks -> pred_masks
inti ln with layer_norm_eps
add_decomposed_rel_pos doc
forward doc of SamForMaskGeneration

ArthurZucker committed:
update processor docstring

ArthurZucker committed:
remove image processor import empty

ArthurZucker committed:
update for testing

ArthurZucker committed:
output vision hidden states + clean recomm
also test all iou values

ArthurZucker committed:
fixup

ArthurZucker committed:
fixup

ArthurZucker committed:
remove unused

ArthurZucker committed:
Update src/transformers/models/sam/modeling_sam.py

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

ArthurZucker committed:
Update src/transformers/models/sam/image_processing_sam.py

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

ArthurZucker committed:
nits

ArthurZucker committed:
Merge branch 'add-sam' of https://github.com/ArthurZucker/transformers into add-sam

younesbelkada commented on docs/source/en/model_doc/sam.mdx:
```diff
@@ -0,0 +1,72 @@
+Resources:
+
+- Demo notebook for using the model
+- Demo notebook for running Automatic Mask Generation (AMG) pipeline.
```
Yes sure, I will work on that after addressing the comments

younesbelkada commented on src/transformers/models/sam/convert_sam_original_to_pytorch.py:
```diff
@@ -0,0 +1,199 @@
+
+
+def convert_sam_checkpoint(model_name, pytorch_dump_folder, push_to_hub):
+    checkpoint_path = hf_hub_download("ybelkada/segment-anything", f"checkpoints/{model_name}.pth")
```
Awesome yes, will transfer them after an approval

sgugger replied:
Mmmm this is for an original checkpoint it looks lilke. Maybe this function should take a repo ID as argument?

younesbelkada committed:
fix

younesbelkada committed:
fix CI tests and slow tests

younesbelkada committed:
replace with Amy's processor

younesbelkada committed:
clearer docstring

younesbelkada committed:
add `SamVisionNeck`

younesbelkada commented on src/transformers/models/sam/modeling_sam.py:
```diff
@@ -0,0 +1,1337 @@
+        if output_hidden_states:
+            all_hidden_states = all_hidden_states + (hidden_states,)
+
+        # Should neck have nn module of its own?
```
Addressed in [93e1364](https://github.com/huggingface/transformers/pull/22654/commits/93e1364faa182521fbe25641fd0fda69e4e1746b)

younesbelkada committed:
refactor - all CI tests should pass

younesbelkada committed:
fix broken import on Gcolab

younesbelkada committed:
few fixes here and there

younesbelkada committed:
fix another bug

sgugger commented on docs/source/en/model_doc/sam.mdx:
```diff
@@ -0,0 +1,74 @@
+
+Resources:
+
+- Demo notebook for using the model
+- Demo notebook for running Automatic Mask Generation (AMG) pipeline. (Will be added once the pipeline is also merged)
```
Links would be nice here (and if not available comment out the line for now).

younesbelkada replied:
Will add a link of the notebook I am working on, and we can comment the pipeline link for now until the other PR gets merged

sgugger commented on docs/source/en/model_doc/sam.mdx:
```diff
@@ -0,0 +1,74 @@
+This model was contributed by [ybelkada](https://huggingface.co/ybelkada) and [ArthurZ](https://huggingface.co/ArthurZ).
+The original code can be found [here](https://github.com/facebookresearch/segment-anything).
+
+Resources:
```
Can you add an example of use or some tips?

younesbelkada replied:
Done in [d71cabf](https://github.com/huggingface/transformers/pull/22654/commits/d71cabf928e9587371f2a5818a39f486a338c649)

sgugger commented on src/transformers/image_processing_utils.py:
```diff
@@ -476,7 +476,10 @@ def is_valid_size_dict(size_dict):
 
 def convert_to_size_dict(
-    size, max_size: Optional[int] = None, default_to_square: bool = True, height_width_order: bool = True
+    size: Optional[int] = None,
```
Why add a wrong type annotation here? The size can be an int, a tuple, a list. Best to leave as it was IMO.

younesbelkada replied:
Reverted in bbcbad1

sgugger commented on src/transformers/models/sam/configuration_sam.py:
```diff
@@ -0,0 +1,344 @@
+logger = logging.get_logger(__name__)
+
+SAM_PRETRAINED_CONFIG_ARCHIVE_MAP = {
+    "facebook/sam-vit-h": "https://huggingface.co/facebook/sam-vit-h/resolve/main/config.json",
+    "facebook/sam-vit-l": "https://huggingface.co/facebook/sam-vit-l/resolve/main/config.json",
+    "facebook/sam-vit-b": "https://huggingface.co/facebook/sam-vit-b/resolve/main/config.json",
```
Needs an update to the new names!

ArthurZucker replied:
They use base for huge so lets call it big

sgugger commented on src/transformers/models/sam/convert_sam_original_to_pytorch.py:
```diff
@@ -0,0 +1,199 @@
+# coding=utf-8
```
This should be named `to_hf_format` as it's not converting anything to PyTorch (the model was already in PyTorch).

sgugger commented on src/transformers/models/sam/image_processing_sam.py:
```diff
@@ -0,0 +1,401 @@
+        """
+        Remove padding and upscale masks to the original image size.
+
+        Args:
+            masks (`torch.Tensor`):
+                Batched masks from the mask_decoder in BxCxHxW format.
+            original_sizes (`torch.Tensor`):
+                The original size of the images before resizing for input to the model, in (H, W) format.
+            reshaped_input_sizes (`torch.Tensor`):
+                The size of the image input to the model, in (H, W) format. Used to remove padding.
+            mask_threshold (`float`, *optional*, defaults to 0.0):
+                The threshold to use for binarizing the masks.
+            binarize (`bool`, *optional*, defaults to `True`):
+                Whether to binarize the masks.
+            pad_size (`int`, *optional*, defaults to `self.pad_size`):
+                The target size the images were padded to before being passed to the model. If None, the target size is
+                assumed to be the processor's `pad_size`.
+        Returns:
+            (`torch.Tensor`): Batched masks in BxCxHxW format, where (H, W) is given by original_size.
```
All B C H W should be spelled out (like we do when documenting inputs of the model).

younesbelkada replied:
Fixed in [4000130](https://github.com/huggingface/transformers/pull/22654/commits/4000130db46cd2c5b5e048d02f5d631d76d05636)

sgugger commented on src/transformers/models/sam/modeling_sam.py:
```diff
@@ -0,0 +1,1359 @@
+_CHECKPOINT_FOR_DOC = "facebook/sam-vit-huge"
+
+SAM_PRETRAINED_MODEL_ARCHIVE_LIST = [
+    "facebook/sam-vit-huge",
+    "facebook/sam-vit-large",
+    "facebook/sam-vit-base",
```
Better, thanks!

sgugger reviewed (commented):
Thanks for iterating. re-the absence of `SamModel`, maybe it's just that `SamForMaskGeneration` needs to be named `SamModel` since there is no other model with a different head? Os should there be a `SamModel` that groups the results of the vision and prompt embedding?

In any case there needs to be something in the mapping for auto-model.

younesbelkada committed:
fix more bugs

ArthurZucker committed:
update and merge

ArthurZucker committed:
Merge branch 'add-sam' of https://github.com/ArthurZucker/transformers into add-sam

younesbelkada committed:
correct ckpt

younesbelkada committed:
address comments

younesbelkada committed:
add tips

younesbelkada committed:
revert

younesbelkada committed:
fix docstring

younesbelkada committed:
replace with `SamModel`

younesbelkada committed:
make fixup

ArthurZucker committed:
add support for bathed images and batch ed points

ArthurZucker committed:
Merge branch 'add-sam' of https://github.com/ArthurZucker/transformers into add-sam

younesbelkada committed:
make fixup this time, really

younesbelkada committed:
Merge branch 'add-sam' of https://github.com/ArthurZucker/transformers into HEAD

younesbelkada committed:
make fixup again and again

younesbelkada committed:
few fixes here and there, this should be the touche finale

sgugger commented on docs/source/en/model_doc/sam.mdx:
```diff
@@ -0,0 +1,97 @@
+
+img_url = "https://huggingface.co/ybelkada/segment-anything/resolve/main/assets/car.png"
+raw_image = Image.open(requests.get(img_url, stream=True).raw).convert("RGB")
+input_points = [[[450, 600]]]
```
Can you put a comment here saying why this point?

younesbelkada replied:
```suggestion
input_points = [[[450, 600]]] # 2D location of a window in the image
```

younesbelkada committed:
Update docs/source/en/model_doc/sam.mdx

younesbelkada committed:
fixup

younesbelkada committed:
correct checkpoints

younesbelkada committed:
correct name

younesbelkada committed:
rm unneeded file

younesbelkada committed:
add notebook

ArthurZucker commented:
Merging as I need it to update the pipeline based on reviews. Will adresse remaining comments in a follow up PR
